File: 120-triangle/120-triangle.py
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        dp = [[0] * i for i in range(1, len(triangle)+1)]
        
        for i in range(len(triangle)):
            if i == 0: dp[0] = triangle[0]; continue
            for j in range(len(triangle[i])):
                left = dp[i-1][j-1] if j > 0 else math.inf
                right = dp[i-1][j] if j < len(triangle[i-1]) else math.inf
                dp[i][j] = triangle[i][j] + min(left, right)
                
        return min(dp[-1])
        
         
#         memo, n = {}, len(triangle)
        
#         def dfs(i,j):
#             if i == n: return 0
#             if (i,j) in memo:
#                 return memo[(i,j)]
            
#             memo[(i,j)] = triangle[i][j] + min(dfs(i+1,j) , dfs(i+1,j+1))
#             return memo[(i,j)]
        
#         return dfs(0,0)

File: 120-triangle/120-triangle.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        int n = triangle.size();
        vector<vector<int>>dp(n, vector<int>(n));
        dp[0][0] = triangle[0][0];
        
        for (int i = 1; i < n; i++) {
            int m = triangle[i-1].size();
            for (int j = 0; j < triangle[i].size(); j++) {
                int tp = (j > 0 ? dp[i-1][j-1]: INT_MAX);
                int tr = (j < m ? dp[i-1][j] : INT_MAX);
                dp[i][j] = min(tp, tr) + triangle[i][j];
            }
        }
        return *min_element(dp.back().begin(), dp.back().end());
    }
};